[2.0.x] PID control fix and more LPC1768 ADC filtering

p3p opened the pull request:
This PR introduces even more sample rejection and filtering for the LPC1768 ADCs (increased median filter from 3 to 23 samples) it also fixes the PID control for all non-avr platforms as the frequency was wrong in the PID_dT calculation. 

fix #8855, fix #8743

tcm0116 commented:
@p3p what do you think about increasing `ADC_MEDIAN_FILTER_SIZE` since it sounds like some people were still seeing noise with the MKS SBASE?

p3p commented:
Yea, there is no reason not to increase it further by default

makz27 commented:
Hi,

FYI I tried a longer print with adc filter size set at 23 and it started to have noise after ~30-45 minutes, I'll try tommorow with Smoothieware value (32) and give a feedback

tcm0116 commented:
@makz27 that's odd that it would start having noise after so long of a time period. Did the noise continue once it started?

Use an odd number, such as 31 if you're going to increase the filter value.

makz27 commented:
@tcm0116 yes once it start, it never stop, looks like it's worst with the time and stop because of the cold extrusion protection.

p3p commented:
Well now I'm just confused, noise like originally? or slow oscillations like was shown in the last screenshot, that didn't look like noise so much as the PID control breaking down slightly.. extruding cooling the hot end?. I cant induce any noise into my ADC readings to test and trust me I've done some very bad wiring to try, so my hours of clean prints with this code is not really useful.

makz27 commented:
adc filter size 31 and a new autotune
 
![screenshot from 2018-01-05 18-36-57](https://user-images.githubusercontent.com/8807832/34620868-40e7d46c-f247-11e7-838f-dbb7c5100b78.png)

I'll check tommorow if it's not my cables

tcm0116 commented:
@makz27 the oscillations before it gets unstable make it look like you've got a PID issue going on. Do you have the change to `PID_dT` in temperature.h from this PR applied to your code?

Bergerac56 commented:
It fixes also  #8928

makz27 commented:
@tcm0116 Yep i have the right change 

`#define PID_dT ((OVERSAMPLENR * float(ACTUAL_ADC_SAMPLES)) / TEMP_TIMER_FREQUENCY)`

makz27 commented:
I've removed the cable chain, the graph looks stable after 1hr of print, i'm still with 31 filter size value, i'll try later with the filter size back to 23

![screenshot from 2018-01-06 13-03-58](https://user-images.githubusercontent.com/8807832/34639572-62e2d422-f2e2-11e7-853a-efb23bff26ab.png)

Edit: 3 hours still stable :)

makz27 commented:
I went back to 23 value and it's stable.

Thanks @p3p & @tcm0116

p3p committed:
LPC1768: Increase ADC median filter from 3 values to 23

Clarify the HAL_adc_get_result method to make sure correct values enter the filters

HAL: Fix the PID control loop for non-AVR platforms

p3p commented:
@thinkyhead @Roxy-3D @Bob-the-Kuhn

Can this be merged, the PID control bug affects all non AVR platforms, filters can always be tuned later
([link](https://github.com/p3p/Marlin/pull/1) for discussion and graphs on the filter)

Bergerac56 commented:
@thinkyhead 
Can we humbly ask to merge this fix? Without it, it is nearly impossible to print with RE-ARM (do not reach targeted temp). And modifications have to be done manually at each new bugfix release. Thanks a lot ;)

thinkyhead commented:
Sorry to have missed this. @p3p Maybe this ["Slack"](https://marlinfirmware.slack.com/) thing…?

p3p commented:
I think I would need invited to join that by email?, not entirely sure how slack works

thinkyhead commented:
I'll check and see if that's the case.

thinkyhead commented:
@p3p Send me an email address to invite you with.
Use this form: http://www.thinkyhead.com/contact selecting the "Marlin Firmware" category.

p3p commented:
Ok sent it, depending on how people use github ... you could probably pull emails off the commits to send invites, although admittedly that would be hit or miss.

thinkyhead commented:
> admittedly that would be hit or miss

And certainly not what everyone would want. Better if everyone gets to choose which email they want to use.
